source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -0,0 +1,228 @@
+#include "extensions/common/dynamic_forward_proxy/dns_cache_impl.h"
+
+#include "common/network/utility.h"
+
+// TODO(mattklein123): Move DNS family helpers to a smaller include.
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace DynamicForwardProxy {
+
+// TODO(mattklein123): circuit breakers / maximums on the number of hosts that the cache can
+//                     contain.
+// TODO(mattklein123): stats
+
+DnsCacheImpl::DnsCacheImpl(
+    Event::Dispatcher& main_thread_dispatcher, ThreadLocal::SlotAllocator& tls,
+    const envoy::config::common::dynamic_forward_proxy::v2alpha::DnsCacheConfig& config)
+    : main_thread_dispatcher_(main_thread_dispatcher),
+      dns_lookup_family_(Upstream::getDnsLookupFamilyFromEnum(config.dns_lookup_family())),
+      resolver_(main_thread_dispatcher.createDnsResolver({})), tls_slot_(tls.allocateSlot()),
+      refresh_interval_(PROTOBUF_GET_MS_OR_DEFAULT(config, dns_refresh_rate, 60000)),
+      host_ttl_(PROTOBUF_GET_MS_OR_DEFAULT(config, dns_refresh_rate, 300000)) {
```

Comment:
`host_ttl`?

Reply:
Good catch, thanks. I will add a test that catches this.